skbio/tree/_tree.py

```diff
@@ -3650,168 +4149,540 @@ def update_result():
+                # no warning:
+                # distances[_start : _stop] += (child.length or 0.0) if length else 1
 
-                starts.append(child.__start)
-                stops.append(child.__stop)
+            node._start, node._stop = min(starts), max(stops)
 
-            node.__start, node.__stop = min(starts), max(stops)
-
-            if len(node.children) > 1:
+            if len(children) > 1:
                 update_result()
 
+            for child in children:
+                del child._start
+                del child._stop
+
+        del self._start
+        del self._stop
+
         return DistanceMatrix(result + result.T, [n.name for n in tip_order])
 
-    def compare_rfd(self, other, proportion=False):
-        """Calculate the Robinson and Foulds symmetric difference.
+    def _compare_topology(
+        self,
+        other,
+        method="subsets",
+        shared_only=True,
+        proportion=False,
+        symmetric=True,
+        include_single=False,
+        weighted=False,
+        metric="euclidean",
+    ):
+        r"""Calculate the topological difference between self and other.
+
+        This function calculates the Robinson-Foulds (RF) distance or its derivates.
 
         Parameters
         ----------
         other : TreeNode
-            A tree to compare against
-        proportion : bool
-            Return a proportional difference
+            The other tree to compare with.
+        method : str, optional
+            Subsets or bipartitions.
+        shared_only : bool, optional
+            Refine to shared taxa.
+        proportion : bool, optional
+            Normalize to fraction.
+        symmetric : bool, optional
+            Symmetric difference.
+        include_single : bool, optional
+            Include singletons.
+        weighted : bool, optional
+            Weight by branch length.
+        metric : str or callable, optional
+            Pairwise distance metric.
 
         Returns
         -------
-        float
-            The distance between the trees
+        int or float
+            Difference between self and other.
+
+        See Also
+        --------
+        compare_rfd
+        compare_wrfd
+        compare_subsets
+        compare_biparts
+
+        """
+        topo1, topo2 = getattr(self, method), getattr(other, method)
+        kwargs = dict(include_single=include_single, map_to_length=weighted)
+        if shared_only:
+            set1, set2 = self.subset(), other.subset()
+            n_shared = len(shared := set1 & set2)
+            sets1 = topo1(within=(shared if len(set1) > n_shared else None), **kwargs)
+            sets2 = topo2(within=(shared if len(set2) > n_shared else None), **kwargs)
+        else:
+            sets1, sets2 = topo1(**kwargs), topo2(**kwargs)
+
+        # unweighted (set difference)
+        if not weighted:
+            method = ("symmetric_" if symmetric else "") + "difference"
```

Comment:
I don't think the use of ternary is common in this codebase (or at least previously wasn't), and these can harm readability...

Reply:
Changed this into simple if/else.